### ISSUE TITLE
Multiple docker images with supvisor servers in them usage case

### ISSUE THREAD

**Issue by awesomebytes** [julien6387/supvisors#99]
Hello,

Thanks for this well documented project!

I'm considering using supvisors as a unified view of a set of Docker containers with supervisor servers inside of them. In my use case, I have a few (currently 3) docker containers where each one of them runs a series of processes. These containers are running in the same machine and they use the host network stack (due to internal random port usage dependencies).

I was trying to do a simple test where I am running a supervisor + supvisor on the host (with the aim that this one would have just a couple of jobs to start the docker containers) and a supervisor + supvisor in a container with a group of processes. I expected to be able to see them in the web ui.

However, I'm finding myself running into socket already in use errors (`zmq.error.ZMQError: Address already in use`).

I wonder if I'm just trying to do something that is impossible or I'm just getting the config wrong.

I have/want:
```
Host OS                  --> supervisor + supvisors with default config
                                                   |__ run docker image(s) start script
Docker container 1 --> supervisor + supvisors with default config
                                                   |__ run process(es) inside of container
Docker container 2 ...
Docker container N 
```

I see in your example scenarios that the different supervisor+supvisors config are executed by ssh-ing into another machine and running supervisor. Being in a different machine I understand that they can re-use ports. If I try to use another port, it runs, but then, I do not know how to tell... the web ui part? to find the additional servers?

I'm still testing, but I thought I'd post already in case someone can give me pointers. Thanks for any help you can provide!

**Comment by julien6387** [comment]
Hi,

Thanks for your feedback.

Supvisors has been designed to 'group' multiple Supervisor instances, all running on a different host but using a set of common parameters (more particularly `inet_http_server.port`, `address_list`, `internal_port`).

It has not been designed to group multiple Supervisor instances running on the same host, so using a quite different configuration, which seems to be your use case.

I'm afraid that there's no short-term solution for you. However, I'm quite interested in such an enhancement and I may consider this in the next months.

At the moment, each item of the `address_list` option is just a host name or IP address.
A preliminary idea would be to define the items as `node_name:inet_http_server.port:internal_port`.
The node key inside Supvisors would thus be based on `(node_name, inet_http_server.port)`, rather than `node_name` only.
Although it seems to be a quite big change, I don't believe that it would be very difficult to implement.

**Comment by awesomebytes** [comment]
@julien6387 if you can point me out to, roughly, which elements would need to be extended to accept `node_name:inet_http_server.port:internal_port` as the input for the node, I could take a look if I am able to provide a proposal.

**Comment by julien6387** [comment]
Hi.

I've just pushed a draft version in the branch dev-0.11.
I've still some work to do (unitary testing + documentation) but integration tests look all right so far.

The `supvisors_list` option (formerly `address_list`, still maintained for now) of the `rpcinterface:supvisors` section must now match elements that identify the Supervisor instances:
`<identifier>host_name:http_port:internal_port`

`identifier` and `:http_port:internal_port` are optional. If not set, Supvisors works as previously.
This is the "regular" case (tested ok):
 * one Supervisor instance per host ;
 * all Supervisor HTTP ports use the same number (taken from SUPERVISOR_SERVER_URL) ;
 * all Supvisors internal ports use the same number (taken from the `internal_port` of the `rpcinterface:supvisors` section) ;
 * the rules file targets host names.

As soon as there is more than one Supervisor instance per host, their HTTP port are different and thus `http_port` must be set. In this case, the rules file targets `host_name:http_port` (not tested yet).

If set, the `identifier` must correspond to the Supervisor `identifier` set in the `supervisord` section of the configuration file.
It must be unique across all Supervisor instances. In this case, the rules file targets Supervisor identifiers, which is a bit more user-friendly (and tested ok).

The use case 'Scenario 3' has been updated to show an example with 6 Supervisor instances: 3 hosts having 2 Supervisor instances each.

Please note that setting the identifier through the supervisord command line (-i identifier) does not work currently. I've proposed a simple pull request to the Supervisor team[#1483](https://github.com/Supervisor/supervisor/pull/1483).

I'd be happy to get your feedback.
If you have any question, please let me know.

Cheers.

**Comment by mnaberez** [comment]
> Please note that setting the identifier through the supervisord command line (-i identifier) does not work currently. I've proposed a simple pull request to the Supervisor team#1483.

[Supervisor 4.2.4](https://pypi.org/project/supervisor/4.2.4/#id1)

**Comment by julien6387** [comment]
Available in Supvisors 0.11

**Comment by awesomebytes** [comment]
Thanks a lot for your excellent work @julien6387 !
I was on holidays (happy 2022!). I'll take a look soon.
